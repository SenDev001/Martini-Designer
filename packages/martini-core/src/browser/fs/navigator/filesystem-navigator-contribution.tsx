import URI from "@theia/core/lib/common/uri";
import { inject, injectable, postConstruct } from "inversify";
import * as React from "react";
import { codeDirResourceRegExp, withoutFileExtension } from "../../../common/fs/file-util";
import { Directory, MartiniFileSystem, Resource } from "../../../common/fs/martini-filesystem";
import { PartialMartiniPackage } from "../../../common/package/martini-package-manager";
import {
    NavigatorContentProviderContribution,
    NavigatorLabelProviderContribution,
    NavigatorOpenHandler
} from "../../navigator/martini-navigator-view-widget";
import {
    AbstractTreeContentProvider,
    AbstractTreeLabelProvider,
    OverlayIcon,
    RefreshContentEvent
} from "../../tree/base-tree";
import { MartiniFileEventDispatcher } from "../martini-filesystem-event-dispatcher";
import { FilesystemPreferences } from "../pref/filesystem-preferences";

@injectable()
export class FileSystemNavigatorContentProvider
    extends AbstractTreeContentProvider
    implements NavigatorContentProviderContribution {
    @inject(MartiniFileSystem)
    private readonly fileSystem: MartiniFileSystem;
    @inject(MartiniFileEventDispatcher)
    private readonly fileEventDispatcher: MartiniFileEventDispatcher;

    readonly priority = 0;
    flattenCodeDirectories: boolean = true;

    @postConstruct()
    init() {
        this.toDispose.push(this.fileEventDispatcher.onFileEvent(e => {
            if (e.action !== "MODIFIED") {
                this.onDidUpdateEmitter.fire(
                    (element: any) =>
                        Resource.is(element) &&
                        (e.paths.find(path =>
                            path.startsWith(element.location)
                        ) as RefreshContentEvent)
                );
            }
        }));
    }

    canHandle(element: any): boolean {
        return (
            PartialMartiniPackage.is(element) ||
            Directory.is(element) ||
            FlattenDirectory.is(element)
        );
    }

    async resolveChildren(parent: any): Promise<any[]> {
        if (PartialMartiniPackage.is(parent)) {
            const packageName = parent.name;
            const resource = await this.fileSystem.get(`/${packageName}`);
            if (Directory.is(resource)) return resource.children.sort((r1, r2) => {
                if (r1.name === "code" && r2.name !== "code") return -1;
                if (r1.name !== "code" && r2.name === "code") return 1;
                return r1.name.localeCompare(r2.name);
            });
        }

        if (FlattenDirectory.is(parent))
            return await this.resolveChildren(
                parent.directories[parent.directories.length - 1]
            );

        if (Directory.is(parent)) {
            const resource = await this.fileSystem.get(parent.location);
            if (Directory.is(resource)) {
                if (this.flattenCodeDirectories) {
                    return Promise.all(
                        this.getSortedChildren(resource).map(async child => {
                            if (Directory.is(child)) {
                                let next = (await this.fileSystem.get(
                                    child.location
                                )) as Resource;
                                const dirs = [];
                                while (
                                    Directory.is(next) &&
                                    !Directory.containsFiles(next) &&
                                    next.children.length === 1
                                ) {
                                    dirs.push(next);
                                    next = (await this.fileSystem.get(
                                        next.children[0].location
                                    )) as Directory;
                                }

                                dirs.push(next);

                                if (dirs.length > 1)
                                    return { ...dirs[dirs.length - 1], directories: dirs } as FlattenDirectory;

                                return child;
                            }
                            return child;
                        })
                    );
                } else if (Directory.is(resource))
                    return this.getSortedChildren(resource);
            }
        }

        return [];
    }

    hasChildren(parent: any): boolean {
        if (PartialMartiniPackage.is(parent)) return parent.status !== "UNLOADED";
        return Directory.is(parent) || FlattenDirectory.is(parent);
    }

    getParent(element: any) {
    }

    private getSortedChildren(directory: Directory) {
        return directory.children.sort((r1, r2) => {
            if ((r1.directory && r2.directory) || (!r1.directory && !r2.directory))
                return r1.name.localeCompare(r2.name);

            return r1.directory ? -1 : 1;
        });
    }
}

@injectable()
export class FileSystemNavigatorLabelProvider extends AbstractTreeLabelProvider
    implements NavigatorLabelProviderContribution {
    priority = 0;

    @inject(FilesystemPreferences)
    private readonly fsPref: FilesystemPreferences;

    @postConstruct()
    protected init() {
        this.toDispose.push(this.fsPref.onPreferenceChanged(e => {
            if (e.preferenceName === "navigator.hideFileExtensions")
                this.onDidUpdateEmitter.fire(undefined);
        }));
    }

    canHandle(element: any): boolean {
        return Resource.is(element) || FlattenDirectory.is(element);
    }

    getCaption(element: any): string | React.ReactNode[] | undefined {
        if (FlattenDirectory.is(element))
            return element.directories.map((dir, i) => (
                <React.Fragment key={i}>
                    <span
                        title={dir.location}
                        className="martini-code-directory-fragment"
                        key={dir.location}
                    >
                        {dir.name}
                    </span>
                    {i + 1 < element.directories.length && <span key={i}>/</span>}
                </React.Fragment>
            ));
        if (Resource.is(element)) {
            let name = element.name;
            if (element.location.match(/^\/core\/code$/)) name = "models";
            if (this.fsPref["navigator.hideFileExtensions"] && !element.directory && codeDirResourceRegExp.test(element.location))
                name = withoutFileExtension(name);

            return [
                <span key="resource-name" title={element.location}>
                    {name}
                </span>
            ];
        }
    }

    getIconClass(element: any): string | undefined {
        if (FlattenDirectory.is(element))
            return this.getIconClass(
                element.directories[element.directories.length - 1]
            );
        if (Directory.is(element)) {
            if (element.location.match(/\/[\w\d_-]{1,50}\/code\/.+/))
                return "martini-tree-icon martini-code-directory-icon";
            return "martini-tree-icon martini-directory-icon";
        }
        if (Resource.isFile(element)) {
            if (element.location.endsWith(".groovy"))
                return "martini-tree-icon martini-groovy-icon";
            return "martini-tree-icon martini-file-icon";
        }
    }

    getOverlayIcons(element: any): OverlayIcon[] | undefined {
        if (
            Directory.is(element) &&
            element.location.match(/^\/[\w\d_-]{1,50}\/code$/)
        )
            return [
                {
                    width: 7,
                    height: 8,
                    iconClass: "martini-code-directory-overlay-icon",
                    position: "top-right"
                }
            ];
        if (
            Directory.is(element) &&
            element.location.match(/\/core\/queries$/)
        )
            return [
                {
                    width: 7,
                    height: 8,
                    iconClass: "martini-query-directory-overlay-icon",
                    position: "top-right"
                }
            ];
    }
}

export interface FlattenDirectory extends Directory {
    directories: Directory[];
}

export namespace FlattenDirectory {
    export function is(object: any): object is FlattenDirectory {
        return !!object && typeof object === "object" && "directories" in object;
    }
}

@injectable()
export class FilesystemNavigatorOpenHandler implements NavigatorOpenHandler {
    getUri(element: any): URI {
        const resource = element as Resource;
        return new URI("martini:/" + resource.location);
    }
    canHandle(element: any): boolean {
        return Resource.is(element) && !element.directory;
    }
}
