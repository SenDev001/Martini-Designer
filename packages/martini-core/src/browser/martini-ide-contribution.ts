import {
    Command,
    CommandContribution,
    CommandRegistry,
    MAIN_MENU_BAR,
    MenuContribution,
    MenuModelRegistry
} from "@theia/core";
import {
    ApplicationShell,
    CommonCommands,
    KeybindingContribution,
    KeybindingRegistry,
    OpenerService
} from "@theia/core/lib/browser";
import { inject, injectable } from "inversify";
import messages from "martini-messages/lib/messages";
import { HistoryManagerProvider } from "./history/history-manager";
import URI from "@theia/core/lib/common/uri";

export const DeleteCommand: Command = {
    id: "martini.delete",
    label: messages.delete,
    iconClass: "martini-icon martini-delete-icon"
};

export const RenameCommand: Command = {
    id: "martini.rename",
    label: messages.rename,
    iconClass: "martini-icon martini-edit-icon"
};

export const EditCommand: Command = {
    id: "martini.edit",
    label: messages.edit,
    iconClass: "martini-icon martini-edit-icon"
};

export const CopyPathCommand: Command = {
    id: "martini.copyPath",
    label: messages.copy_path,
    iconClass: "martini-icon martini-copy-icon"
};

export const CopyNamespaceCommand: Command = {
    id: "martini.copyNamespace",
    label: messages.copy_namespace,
    iconClass: "martini-icon martini-copy-icon"
};

export type MoveDirection = "left" | "right" | "up" | "down";

export const MoveLeftCommand: Command = {
    id: "martini.moveLeft",
    label: messages.move_left,
    iconClass: "martini-icon martini-arrow-left-icon"
};

export const MoveRightCommand: Command = {
    id: "martini.moveRight",
    label: messages.move_right,
    iconClass: "martini-icon martini-arrow-right-icon"
};

export const MoveUpCommand: Command = {
    id: "martini.moveUp",
    label: messages.move_up,
    iconClass: "martini-icon martini-arrow-up-icon"
};

export const MoveDownCommand: Command = {
    id: "martini.moveDown",
    label: messages.move_down,
    iconClass: "martini-icon martini-arrow-down-icon"
};

export const EditMartiniPropertiesCommand: Command = {
    id: "martini.editMartiniProperties",
    label: messages.edit_martini_properties,
    iconClass: "martini-icon martini-edit-icon"
};

export const EditComments: Command = {
    id: "martini.editComments",
    label: messages.edit_comments,
    iconClass: "martini-icon martini-comments-icon"
};

export const ToggleEnabled: Command = {
    id: "martini.toggleEnabled",
    label: messages.toggle_enabled,
    iconClass: "martini-icon martini-disabled-icon"
};

export namespace MartiniMenus {
    export const MARTINI = [...MAIN_MENU_BAR, "0_martini"];
}

@injectable()
export class MartiniIdeCommandContribution implements CommandContribution {
    @inject(ApplicationShell)
    private readonly shell: ApplicationShell;
    @inject(OpenerService)
    private readonly openerService: OpenerService;

    registerCommands(commands: CommandRegistry): void {
        commands.registerCommand(DeleteCommand);
        commands.registerCommand(RenameCommand);
        commands.registerCommand(CopyPathCommand);
        commands.registerCommand(CopyNamespaceCommand);
        commands.registerCommand(MoveLeftCommand);
        commands.registerCommand(MoveRightCommand);
        commands.registerCommand(MoveUpCommand);
        commands.registerCommand(MoveDownCommand);
        commands.registerCommand(EditComments);
        commands.registerCommand(ToggleEnabled);
        commands.registerCommand(EditCommand);
        CommonCommands.COPY.iconClass = "martini-icon martini-copy-icon";
        CommonCommands.CUT.iconClass = "martini-icon martini-cut-icon";
        CommonCommands.PASTE.iconClass = "martini-icon martini-paste-icon";
        CommonCommands.FIND.iconClass = "martini-icon martini-search-icon";

        const isTextElementFocused = () => {
            const activeElement = document.activeElement;
            return (
                (activeElement instanceof HTMLInputElement && activeElement.type === "text") ||
                activeElement instanceof HTMLTextAreaElement
            );
        };
        commands.registerHandler(CommonCommands.UNDO.id, {
            execute: () => {
                document.execCommand("undo", true);
            },
            isEnabled: isTextElementFocused
        });
        commands.registerHandler(CommonCommands.REDO.id, {
            execute: () => {
                document.execCommand("redo", true);
            },
            isEnabled: isTextElementFocused
        });
        commands.registerHandler(CommonCommands.UNDO.id, {
            execute: () => {
                const widget = this.shell.currentWidget;
                return HistoryManagerProvider.is(widget) && widget.historyManager?.undo();
            },
            isEnabled: () => {
                const widget = this.shell.currentWidget;
                return (
                    HistoryManagerProvider.is(widget) &&
                    widget.historyManager !== undefined &&
                    widget.historyManager.canUndo()
                );
            }
        });
        commands.registerHandler(CommonCommands.REDO.id, {
            execute: () => {
                const widget = this.shell.currentWidget;
                return HistoryManagerProvider.is(widget) && widget.historyManager?.redo();
            },
            isEnabled: () => {
                const widget = this.shell.currentWidget;
                return (
                    HistoryManagerProvider.is(widget) &&
                    widget.historyManager !== undefined &&
                    widget.historyManager.canRedo()
                );
            }
        });
        commands.registerCommand(EditMartiniPropertiesCommand, {
            execute: async () => {
                const uri: URI = new URI("martini://core/conf/properties/application.properties");
                const handler = await this.openerService.getOpener(uri);
                return handler.open(uri);
            }
        });
    }
}

@injectable()
export class MartiniMenuContribution implements MenuContribution {
    registerMenus(menus: MenuModelRegistry): void {
        menus.registerSubmenu(MartiniMenus.MARTINI, "Martini");
        menus.registerMenuAction(MartiniMenus.MARTINI, {
            commandId: EditMartiniPropertiesCommand.id
        });
    }
}

@injectable()
export class MartiniIdeKeybindingContribution implements KeybindingContribution {
    registerKeybindings(keybindings: KeybindingRegistry): void {
        keybindings.registerKeybindings(
            {
                command: DeleteCommand.id,
                keybinding: "del"
            },
            {
                command: RenameCommand.id,
                keybinding: "f2"
            },
            {
                command: EditCommand.id,
                keybinding: "ctrlcmd+enter"
            },
            {
                command: CopyPathCommand.id,
                keybinding: "ctrlcmd+shift+C"
            },
            {
                command: CopyNamespaceCommand.id,
                keybinding: "ctrlcmd+alt+shift+C"
            },
            {
                command: CommonCommands.PASTE.id,
                keybinding: "ctrlcmd+V"
            },
            {
                command: MoveLeftCommand.id,
                keybinding: "ctrlcmd+left"
            },
            {
                command: MoveRightCommand.id,
                keybinding: "ctrlcmd+right"
            },
            {
                command: MoveUpCommand.id,
                keybinding: "ctrlcmd+up"
            },
            {
                command: MoveDownCommand.id,
                keybinding: "ctrlcmd+down"
            },
            {
                command: EditComments.id,
                keybinding: "ctrlcmd+/"
            },
            {
                command: ToggleEnabled.id,
                keybinding: "alt+/"
            }
        );
    }
}
