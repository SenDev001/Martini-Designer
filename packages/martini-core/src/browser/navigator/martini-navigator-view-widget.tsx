import { ContributionProvider } from "@theia/core";
import {
    ContextMenuRenderer,
    OpenerService,
    TreeProps
} from "@theia/core/lib/browser";
import URI from "@theia/core/lib/common/uri";
import { inject, injectable, named } from "inversify";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import { MartiniFileSystem } from "../../common/fs/martini-filesystem";
import {
    AbstractTreeContentProvider,
    AbstractTreeLabelProvider,
    BaseTreeModel,
    OverlayIcon,
    TreeContentProvider,
    TreeLabelProvider
} from "../tree/base-tree";
import { BaseTreeWidget, DnDHandler, DragData, DropPosition } from "../tree/base-tree-widget";

export const NavigatorOpenHandler = Symbol("NavigatorOpenHandler");

export interface NavigatorOpenHandler {
    getUri: (element: any) => URI;
    canHandle: (element: any) => boolean;
}

@injectable()
export class Navigator extends BaseTreeWidget {
    static readonly ID = "martini-core:navigator";

    @inject(ContributionProvider)
    @named(NavigatorOpenHandler)
    private openHandlers: ContributionProvider<NavigatorOpenHandler>;

    constructor(
        @inject(TreeContentProvider)
        protected readonly treeContentProvider: TreeContentProvider,
        @inject(TreeLabelProvider)
        protected readonly treeLabelProvider: TreeLabelProvider,
        @inject(OpenerService) readonly openerService: OpenerService,
        @inject(TreeProps) readonly props: TreeProps,
        @inject(BaseTreeModel) readonly model: BaseTreeModel,
        @inject(ContextMenuRenderer)
        protected readonly contextMenuRenderer: ContextMenuRenderer
    ) {
        super(
            treeContentProvider,
            treeLabelProvider,
            props,
            model,
            contextMenuRenderer
        );
        this.id = Navigator.ID;
        this.title.label = messages.navigator;
        this.title.caption = messages.navigator;
        this.title.iconClass = "martini-tab-icon martini-server-icon";
        setTimeout(() => {
            this.model.input = new NavigatorRoot();
            this.update();
        }, 1000);
    }

    handleDoubleClick(element: any): boolean {
        const handler = this.openHandlers.getContributions().find(handler => handler.canHandle(element));
        if (handler) {
            const uri = handler.getUri(element);
            this.openerService.getOpener(uri).then(opener => opener.open(uri));
            return true;
        }

        return false;
    }
}

export const NavigatorContentProviderContribution = Symbol(
    "NavigatorContentProviderContribution"
);

export interface NavigatorContentProviderContribution
    extends TreeContentProvider {
    canHandle(element: any): boolean;
    priority: number;
}

export const NavigatorLabelProviderContribution = Symbol(
    "NavigatorLabelProviderContribution"
);

export interface NavigatorLabelProviderContribution extends TreeLabelProvider {
    canHandle(element: any): boolean;
    priority: number;
}

export class NavigatorRoot { }

@injectable()
export class NavigatorTreeContentProvider extends AbstractTreeContentProvider {
    @inject(MartiniFileSystem)
    fileSystem: MartiniFileSystem;
    readonly contributions: NavigatorContentProviderContribution[];

    constructor(
        @inject(ContributionProvider)
        @named(NavigatorContentProviderContribution)
        contributionProvider: ContributionProvider<
            NavigatorContentProviderContribution
        >
    ) {
        super();
        this.contributions = contributionProvider
            .getContributions()
            .sort((c1, c2) => c2.priority - c1.priority);
        this.contributions.forEach(contrib => {
            if (contrib.onDidUpdate)
                contrib.onDidUpdate(e => this.onDidUpdateEmitter.fire(e));
        });
    }

    async resolveChildren(parent: any): Promise<any[]> {
        const children = [];
        for (const contrib of this.contributions) {
            if (contrib.canHandle(parent)) {
                const result = await contrib.resolveChildren(parent);
                children.push(...result);
            }
        }

        return children;
    }

    hasChildren(parent: any): boolean {
        for (const contrib of this.contributions) {
            if (contrib.hasChildren(parent)) return true;
        }
        if (parent instanceof NavigatorRoot) return true;
        return false;
    }

    getParent(element: any) {
        for (const contrib of this.contributions) {
            const result = contrib.getParent(element);
            if (result) return true;
        }

        return undefined;
    }

    isEquals(element: any, other: any) {
        if (element instanceof NavigatorRoot && other instanceof NavigatorRoot)
            return true;
        for (const contrib of this.contributions) {
            if (contrib.isEqual) {
                const result = contrib.isEqual(element, other);
                if (result) return true;
            }
        }
        return false;
    }
}

@injectable()
export class NavigatorTreeLabelProvider extends AbstractTreeLabelProvider {
    readonly contributions: NavigatorLabelProviderContribution[];

    constructor(
        @inject(ContributionProvider)
        @named(NavigatorLabelProviderContribution)
        contributionProvider: ContributionProvider<
            NavigatorLabelProviderContribution
        >
    ) {
        super();
        this.contributions = contributionProvider
            .getContributions()
            .sort((c1, c2) => c2.priority - c1.priority);
        this.contributions.forEach(contrib => {
            if (contrib.onDidUpdate)
                contrib.onDidUpdate(e => this.onDidUpdateEmitter.fire(e));
        });
    }

    getCaption(element: any): string | React.ReactNode[] | undefined {
        for (const contrib of this.contributions) {
            if (contrib.canHandle(element)) {
                const caption = contrib.getCaption(element);
                if (caption) return caption;
            }
        }

        return undefined;
    }

    getIconClass(element: any): string | undefined {
        for (const contrib of this.contributions) {
            if (contrib.getIconClass && contrib.canHandle(element)) {
                const iconClass = contrib.getIconClass(element);
                if (iconClass) return iconClass;
            }
        }
    }

    getOverlayIcons(element: any): OverlayIcon[] | undefined {
        const overlays: OverlayIcon[] = [];
        for (const contrib of this.contributions) {
            if (contrib.getOverlayIcons && contrib.canHandle(element)) {
                const icons = contrib.getOverlayIcons(element);
                if (icons) overlays.push(...icons);
            }
        }

        return overlays;
    }
}

@injectable()
export class NavigatorDnDHandler implements DnDHandler {

    isDraggable(_element: any): boolean {
        return true;
    }

    getDragData(_elements: any[]): DragData[] {
        return [];
    }

    async handleDrop(event: React.DragEvent<Element>, _target: any, _position: DropPosition): Promise<void> {
        event.dataTransfer.dropEffect = "none";
    }

}
