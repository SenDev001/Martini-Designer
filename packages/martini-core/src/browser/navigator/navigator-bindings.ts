import { bindContributionProvider } from "@theia/core";
import { bindViewContribution, defaultTreeProps, WidgetFactory } from "@theia/core/lib/browser";
import { TabBarToolbarContribution } from "@theia/core/lib/browser/shell/tab-bar-toolbar";
import { interfaces } from "inversify";
import createBaseTreeContainer from "../tree/base-tree-container";
import { DnDHandler } from "../tree/base-tree-widget";
import { NavigatorViewContribution, NAVIGATOR_VIEW_CONTEXT_MENU } from "./martini-navigator-view-contribution";
import {
    Navigator,
    NavigatorContentProviderContribution,
    NavigatorDnDHandler,
    NavigatorLabelProviderContribution,
    NavigatorOpenHandler,
    NavigatorTreeContentProvider,
    NavigatorTreeLabelProvider
} from "./martini-navigator-view-widget";

export const bindNavigatorBindings = (bind: interfaces.Bind) => {
    bindContributionProvider(bind, NavigatorContentProviderContribution);
    bindContributionProvider(bind, NavigatorLabelProviderContribution);
    bindContributionProvider(bind, NavigatorOpenHandler);
    bindViewContribution(bind, NavigatorViewContribution);
    bind(TabBarToolbarContribution).toService(NavigatorViewContribution);
    bind(Navigator).toSelf();
    bind(NavigatorTreeLabelProvider).toSelf();
    bind(WidgetFactory).toDynamicValue(ctx => ({
        id: Navigator.ID,
        createWidget: () => {
            const child = createBaseTreeContainer(
                ctx.container,
                {
                    contentProvider: NavigatorTreeContentProvider,
                    labelProvider: NavigatorTreeLabelProvider
                },
                {
                    ...defaultTreeProps,
                    multiSelect: true,
                    search: false,
                    contextMenuPath: NAVIGATOR_VIEW_CONTEXT_MENU,
                    globalSelection: true
                }
            );
            child.bind(DnDHandler).to(NavigatorDnDHandler);
            child.bind(Navigator).toSelf();
            return child.get(Navigator);
        }
    }));
};
