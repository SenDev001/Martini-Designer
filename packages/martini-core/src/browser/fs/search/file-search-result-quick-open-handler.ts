import { OpenerService, WidgetManager } from "@theia/core/lib/browser";
import { QuickOpenItemOptions, QuickOpenMode } from "@theia/core/lib/common/quick-open-model";
import URI from "@theia/core/lib/common/uri";
import { inject, injectable } from "inversify";
import { Directory, Resource } from "../../../common/fs/martini-filesystem";
import { PartialMartiniPackage } from "../../../common/package/martini-package-manager";
import { FileSearchResult, ResourceSearchResult } from "../../../common/search/resource-search-service";
import { Navigator, NavigatorTreeLabelProvider } from "../../navigator/martini-navigator-view-widget";
import { ResourceSearchResultQuickOpenHandler } from "../../search/resource-search-quick-open";

@injectable()
export class FileSearchResultQuickOpenHandler implements ResourceSearchResultQuickOpenHandler {
    @inject(OpenerService)
    private readonly openerService: OpenerService;
    @inject(NavigatorTreeLabelProvider)
    private readonly labelProvider: NavigatorTreeLabelProvider;
    @inject(WidgetManager)
    private readonly widgetMngr: WidgetManager;

    supports(result: ResourceSearchResult): boolean {
        return FileSearchResult.is(result);
    }

    getOptions(result: FileSearchResult): QuickOpenItemOptions {
        const resource: Resource | Directory = {
            location: result.location,
            name: result.name,
            directory: result.directory,
            lastModified: result.lastModified,
            readOnly: false,
            size: 0,
            children: result.directory ? [] : undefined
        };

        return {
            label: result.location,
            iconClass: this.labelProvider.getIconClass(resource),
            run: mode => {
                if (mode === QuickOpenMode.OPEN || mode === QuickOpenMode.OPEN_IN_BACKGROUND) return this.open(result);
                return false;
            }
        };
    }

    private open(result: ResourceSearchResult): boolean {
        let uri: URI | undefined;
        if (FileSearchResult.is(result)) {
            if (!result.directory) uri = new URI(`martini:/${result.location}`);
            else {
                this.openDir(result);
                return true;
            }
        }
        if (uri) {
            this.openerService.getOpener(uri).then(opener => opener.open(uri!));
            return true;
        }

        return false;
    }

    private async openDir(result: FileSearchResult): Promise<void> {
        const navigator = (await this.widgetMngr.getOrCreateWidget(Navigator.ID)) as Navigator;
        await navigator.model.expandElements(element => {
            if (PartialMartiniPackage.is(element)) return result.location.startsWith(element.name, 1);

            return Resource.is(element) && element.directory && result.location.startsWith(element.location);
        });
        await navigator.model.selectElements(
            element => Resource.is(element) && element.directory && result.location === element.location
        );
        navigator.activate();
    }
}
