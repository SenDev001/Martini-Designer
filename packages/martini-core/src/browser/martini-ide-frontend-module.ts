import { CommandContribution, ILogger, LogLevel, MenuContribution } from "@theia/core";
import {
    FrontendApplicationContribution,
    KeybindingContribution,
    WebSocketConnectionProvider
} from "@theia/core/lib/browser";
import { ContainerModule, inject, injectable, interfaces } from "inversify";
import "../../src/browser/style/index.css";
import { MartiniAccountManager, martiniAccountManagerPath } from "../common/instance/martini-account-manager";
import {
    MartiniInstanceInfoManager,
    martiniInstanceInfoManagerPath
} from "../common/instance/martini-instance-info-manager";
import { ResourceSearchService, resourceSearchServicePath } from "../common/search/resource-search-service";
import { ServiceSearchService, serviceSearchServicePath } from "../common/search/service-search-service";
import { DocumentTypeManager, documentTypeManagerPath } from "../common/tracker/document-type-manager";
import { ContentAssistContributionManager } from "./content-assist/content-assist-contribution";
import { bindDatabaseConnectionBindings } from "./database-connection/database-connection-bindings";
import { ApiResponseHandler } from "./dialogs/api-response-handler";
import { LocalDnDTransfer } from "./dnd/local-dnd-transfer";
import { bindEmbeddedTextEditor } from "./editor/embedded-text-editor";
import { bindMarkdownEditorDialog } from "./editor/md/markdown-editor";
import { bindEndpointBindings } from "./endpoint/endpoint-bindings";
import { bindFilesystemBindings } from "./fs/filesystem-bindings";
import { bindJmsBindings } from "./jms/jms-bindings";
import {
    MartiniIdeCommandContribution,
    MartiniIdeKeybindingContribution,
    MartiniMenuContribution
} from "./martini-ide-contribution";
import { bindNavigatorBindings } from "./navigator/navigator-bindings";
import { bindMartiniPackageBindings } from "./package/martini-package-bindings";
import { DefaultProgressService, ProgressService } from "./progress/progress-service";
import { bindPropertiesViewBindings } from "./properties-view/properties-view-bindings";
import { bindResourceSearchBindings } from "./search/resource-search-bindings";
import { bindWizardBindings } from "./wizard/wizard-bindings";
import { initYupExt } from "./yup/yup-ext";
import { initYupLocale } from "./yup/yup-locale";

export const configureModule: interfaces.ContainerModuleCallBack = (bind, unbind, isBound, rebind) => {
    bindPropertiesViewBindings(bind);
    bindNavigatorBindings(bind);
    bindMartiniPackageBindings(bind);
    bindEndpointBindings(bind);
    bindDatabaseConnectionBindings(bind);
    bindFilesystemBindings(bind);
    bindWizardBindings(bind);
    bindEmbeddedTextEditor(bind, rebind);
    bindMarkdownEditorDialog(bind);
    bindJmsBindings(bind);
    bind(MenuContribution)
        .to(MartiniMenuContribution)
        .inSingletonScope();
    bind(CommandContribution)
        .to(MartiniIdeCommandContribution)
        .inSingletonScope();
    bind(KeybindingContribution)
        .to(MartiniIdeKeybindingContribution)
        .inSingletonScope();
    bind(FrontendApplicationContribution).to(MartiniIdeFrontendContribution);
    bind(ProgressService)
        .to(DefaultProgressService)
        .inSingletonScope();
    bind(MartiniInstanceInfoManager)
        .toDynamicValue(ctx => {
            const connection = ctx.container.get(WebSocketConnectionProvider);
            return connection.createProxy<MartiniInstanceInfoManager>(martiniInstanceInfoManagerPath);
        })
        .inSingletonScope();
    bind(MartiniAccountManager)
        .toDynamicValue(ctx => {
            const connection = ctx.container.get(WebSocketConnectionProvider);
            return connection.createProxy<MartiniAccountManager>(martiniAccountManagerPath);
        })
        .inSingletonScope();
    bind(DocumentTypeManager)
        .toDynamicValue(ctx => {
            const connection = ctx.container.get(WebSocketConnectionProvider);
            return connection.createProxy(documentTypeManagerPath);
        })
        .inSingletonScope();
    bind(ResourceSearchService)
        .toDynamicValue(ctx => {
            const connection = ctx.container.get(WebSocketConnectionProvider);
            return connection.createProxy(resourceSearchServicePath);
        })
        .inSingletonScope();
    bind(ServiceSearchService)
        .toDynamicValue(ctx => {
            const connection = ctx.container.get(WebSocketConnectionProvider);
            return connection.createProxy(serviceSearchServicePath);
        })
        .inSingletonScope();
    bind(ApiResponseHandler)
        .toSelf()
        .inSingletonScope();
    bind(LocalDnDTransfer).toConstantValue(LocalDnDTransfer.INSTANCE);
    bindResourceSearchBindings(bind);
    bind(ContentAssistContributionManager)
        .toSelf()
        .inSingletonScope();
};

export default new ContainerModule(configureModule);

@injectable()
export class MartiniIdeFrontendContribution implements FrontendApplicationContribution {
    @inject(MartiniInstanceInfoManager)
    instanceInfoManager: MartiniInstanceInfoManager;
    @inject(MartiniAccountManager)
    accountManager: MartiniAccountManager;
    @inject(ILogger)
    private readonly logger: ILogger;

    initialize() {
        this.logger.setLogLevel(LogLevel.INFO);
        initYupExt();
        initYupLocale();
    }
}
