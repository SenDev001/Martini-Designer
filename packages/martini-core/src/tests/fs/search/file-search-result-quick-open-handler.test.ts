require("reflect-metadata");
import { OpenerService, OpenHandler, QuickOpenMode, WidgetManager } from "@theia/core/lib/browser";
import URI from "@theia/core/lib/common/uri";
import { Container } from "inversify";
import { FileSearchResultQuickOpenHandler } from "../../../browser/fs/search/file-search-result-quick-open-handler";
import { NavigatorTreeLabelProvider } from "../../../browser/navigator/martini-navigator-view-widget";

jest.mock("../../../browser/navigator/martini-navigator-view-widget");
jest.mock("@theia/core/lib/browser/widget-manager");

const container = new Container();
//@ts-ignore
container.bind(NavigatorTreeLabelProvider).toConstantValue(new NavigatorTreeLabelProvider());
container.bind(WidgetManager).toConstantValue(new WidgetManager());
container.bind(FileSearchResultQuickOpenHandler).toSelf();
const mockOpenerService: OpenerService = {
    getOpener: jest.fn(),
    getOpeners: jest.fn()
};
container.bind(OpenerService).toConstantValue(mockOpenerService);
const handler = container.get(FileSearchResultQuickOpenHandler);

beforeEach(() => jest.clearAllMocks());

test("Should support FileSearchResult", () => {
    const actual = handler.supports({
        type: "path",
        location: "test"
    });
    expect(actual).toBe(true);
});

test("Should not support unknown SearchResult", () => {
    const actual = handler.supports({
        //@ts-ignore
        type: "test",
        location: "test"
    });
    expect(actual).toBe(false);
});

test("Should return QuickOpenItemOptions", () => {
    const actual = handler.getOptions({
        directory: false,
        lastModified: 0,
        extension: "gloop",
        name: "Test.gloop",
        location: "/test/code/Test.gloop",
        type: "path"
    });
    expect(actual).toMatchSnapshot();
});

test("Should open the file", done => {
    const options = handler.getOptions({
        directory: false,
        lastModified: 0,
        extension: "gloop",
        name: "Test.gloop",
        location: "/test/code/Test.gloop",
        type: "path"
    });
    const mockOpener: OpenHandler = {
        id: "mock",
        canHandle: () => 1,
        open: jest.fn(async () => {
            expect(mockOpenerService.getOpener).toBeCalledTimes(1);
            expect(mockOpenerService.getOpener).toBeCalledWith(new URI("martini://test/code/Test.gloop"));
            expect(mockOpener.open).toBeCalledTimes(1);
            expect(mockOpener.open).toBeCalledWith(new URI("martini://test/code/Test.gloop"));
            done();
        })
    };
    mockOpenerService.getOpener = jest.fn(async () => mockOpener);
    const result = options.run?.(QuickOpenMode.OPEN);
    expect(result).toBe(true);
});
