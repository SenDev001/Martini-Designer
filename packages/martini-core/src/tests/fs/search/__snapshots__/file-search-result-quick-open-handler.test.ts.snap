// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return QuickOpenItemOptions 1`] = `
Object {
  "iconClass": undefined,
  "label": "/test/code/Test.gloop",
  "run": [Function],
}
`;
